src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2055,458 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+	EnsureTableOwner(relationId);
+
+	/* we want to serialize all the metadata changes to this table */
+	LockRelationOid(relationId, ExclusiveLock);
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+
+		/*
+		 * Even if the table owner is a malicious user, as long as the shard placements
+		 * fit into basic requirements of Citus metadata, the user can only affect its
+		 * own tables. Given that the user is owner of the table, we should allow.
+		 */
+		EnsureShardPlacementMetadataIsSane(shardId, placementId, shardState,
+										   shardLength, groupId);
+	}
+
+	InsertShardPlacementRow(shardId, placementId, shardState, shardLength, groupId);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * EnsureShardPlacementMetadataIsSane ensures if the input parameters for
+ * the shard placement metadata is sane.
+ */
+static void
+EnsureShardPlacementMetadataIsSane(int64 shardId, int64 placementId,
+								   int32 shardState, int64 shardLength,
+								   int32 groupId)
+{
+	if (shardId <= INVALID_SHARD_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard id: %ld", shardId)));
+	}
+
+	if (placementId <= INVALID_PLACEMENT_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Shard placement has invalid placement id "
+							   "(%ld) for shard(%ld)", placementId, shardId)));
+	}
+
+	if (!(shardState == SHARD_STATE_ACTIVE ||
+		  shardState == SHARD_STATE_TO_DELETE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard state: %d", shardState)));
+	}
+
+	bool nodeFound = false;
+	bool includeNodesFromOtherClusters = false;
+	List *workerNodeList = ReadDistNode(includeNodesFromOtherClusters);
+	WorkerNode *currentNode = NULL;
+	foreach_ptr(currentNode, workerNodeList)
+	{
+		if (currentNode->groupId == groupId)
+		{
+			nodeFound = true;
+			break;
+		}
+	}
+	if (!nodeFound)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Node with group id %d for shard placement "
+							   "%ld does not exist", groupId, shardId)));
+	}
+}
+
+
+/*
+ * ShouldSkipMetadataChecks returns true if the current user is allowed to
+ * make any
+ */
+static bool
+ShouldSkipMetadataChecks(void)
+{
+	if (strcmp(EnableManualMetadataChangesForUser, "") != 0)
+	{
+		/*
+		 * EnableManualMetadataChangesForUser is a GUC which
+		 * can be changed by a super user. We use this GUC as
+		 * a safety belt in case the current metadata checks are
+		 * too restrictive and the operator can allow users to skip
+		 * the checks.
+		 */
+
+		/*
+		 * Make sure that the user exists, and print it to prevent any
+		 * optimization skipping the get_role_oid call.
+		 */
+		bool missingOK = false;
+		Oid allowedUserId = get_role_oid(EnableManualMetadataChangesForUser, missingOK);
```

Comment:
Do we really want the user to be missing to be an error. I think if we show a warning and return `false` it would be good enough.

Reply:
I think if you are setting this GUC, it'd be good to see the error as the user does not exists?